// bare-cov/lib/transformer.js
/*
 * Copyright (c) 2024 Vinh Vu
 * Email: mrthanhvinh168@gmail.com
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * This source code is the property of Vinh Vu. It cannot be re-used or shared
 * with anyone else without his consent.
 */

'use strict'

const { isAbsolute } = require('path')
const { fileURLToPath } = require('url')
const libCoverage = require('istanbul-lib-coverage')
const defaultExclude = require('@istanbuljs/schema/default-exclude')
const defaultExtension = require('@istanbuljs/schema/default-extension')
const libReport = require('istanbul-lib-report')
const v8ToIstanbul = require('v8-to-istanbul')
const reports = require('istanbul-reports')
const TestExclude = require('./test-exclude')

class Transformer {
  constructor (opts = {}) {
    this.includeRelative = opts.includeRelative ?? false
    this.exclude = new TestExclude({
      exclude: defaultExclude,
      include: [],
      extension: defaultExtension,
      relativePath: true,
      excludeNodeModules: true
    })
    this.dir = opts.dir ?? 'coverage'
    this.reporters = opts.reporters ?? ['text', 'json']
    this.reporterOptions = opts.reporterOptions ?? {}

    this.includedUrlCache = new Map()
  }

  normalizeUrl (v8ReportResult) {
    if (/^node:/.test(v8ReportResult.url)) {
      v8ReportResult.url = `${v8ReportResult.url.replace(/^node:/, '')}.js`
    }

    if (/^file:\/\//.test(v8ReportResult.url)) {
      v8ReportResult.url = fileURLToPath(v8ReportResult.url)
    }

    return v8ReportResult
  }

  isResultUrlIncluded (url) {
    const cacheResult = this.includedUrlCache.get(url)
    if (cacheResult !== undefined) return cacheResult

    const result = (this.includeRelative || isAbsolute(url)) && this.exclude.shouldInstrument(url)
    this.includedUrlCache.set(url, result)
    return result
  }

  async transformToCoverageMap (rawV8Report) {
    const v8Report = {
      result: rawV8Report.result
        .map(v8ReportResult => this.normalizeUrl(v8ReportResult))
        .filter(v8ReportResult => this.isResultUrlIncluded(v8ReportResult.url))
    }

    const coverageMap = libCoverage.createCoverageMap()
    for (const v8ReportResult of v8Report.result) {
      const converter = v8ToIstanbul(v8ReportResult.url)
      await converter.load()
      converter.applyCoverage(v8ReportResult.functions)
      coverageMap.merge(converter.toIstanbul())
    }

    return coverageMap
  }

  report (coverageMap) {
    const context = libReport.createContext({ dir: this.dir, coverageMap })

    for (const reporter of this.reporters) {
      reports.create(reporter, {
        skipEmpty: false,
        skipFull: false,
        maxCols: process.stdout.columns || 100,
        ...this.reporterOptions[reporter]
      }).execute(context)
    }
  }
}

module.exports = Transformer
